File: z3/diet.py
#!/usr/bin/python -u
# -*- coding: latin-1 -*-
# 
# Simple diet problem in in Z3
#
# Standard Operations Research example.
#
# Minimize the cost for the products:
# Type of                        Calories   Chocolate    Sugar    Fat
# Food                                      (ounces)     (ounces) (ounces)
# Chocolate Cake (1 slice)       400           3            2      2
# Chocolate ice cream (1 scoop)  200           2            2      4
# Cola (1 bottle)                150           0            4      1
# Pineapple cheesecake (1 piece) 500           0            4      5
#
#
# This Z3 model was written by Hakan Kjellerstrand (hakank@gmail.com)
# See also my Z3 page: http://hakank.org/z3/
# 
#
from __future__ import print_function
from z3_utils_hakank import *


#
# Diet problem:
#  variant = 1: use direct sums
#  variant = 2: use scalar product
#
def diet(variant=1):

  sol = Optimize()

  # data
  n = 4
  price = [50, 20, 30, 80]  # in cents
  limits = [500, 6, 10, 8]  # requirements for each nutrition type

  # nutritions for each product
  calories  = [400, 200, 150, 500]
  chocolate = [  3,   2,   0,   0]
  sugar     = [  2,   2,   4,   4]
  fat       = [  2,   4,   1,   5]

  #
  # declare variables
  #
  x = makeIntVars(sol, "x",n,0,100)
  cost = makeIntVar(sol,"cost",0, 10000)

  #
  # constraints
  #
  if variant == 1:
      sol.add(Sum([x[i] * calories[i] for i in range(n)]) >= limits[0])
      sol.add(Sum([x[i] * chocolate[i] for i in range(n)]) >= limits[1])
      sol.add(Sum([x[i] * sugar[i] for i in range(n)]) >= limits[2])
      sol.add(Sum([x[i] * fat[i] for i in range(n)]) >= limits[3])
  else:
      sol.add(scalar_product2(sol,x, calories) >= limits[0])
      sol.add(scalar_product2(sol,x, chocolate) >= limits[1])
      sol.add(scalar_product2(sol,x, sugar) >= limits[2])
      sol.add(scalar_product2(sol,x, fat) >= limits[3])
      
  # objective
  sol.minimize(cost)

  # solution

  if sol.check() == sat:
      mod = sol.model()
      print("cost:", mod.eval(cost))
      print([("abcdefghij"[i], mod.eval(x[i])) for i in range(n)])


if __name__ == "__main__":
  diet(1)
  print()
  diet(2)

File: swi_prolog/diet.pl
/*

  Diet problem in SWI Prolog

  Standard diet problem.

  Minimize the cost for the products:
   Type of                        Calories   Chocolate    Sugar    Fat
   Food                                      (ounces)     (ounces) (ounces)
   Chocolate Cake (1 slice)       400           3            2      2
   Chocolate ice cream (1 scoop)  200           2            2      4
   Cola (1 bottle)                150           0            4      1
   Pineapple cheesecake (1 piece) 500           0            4      5



  Model created by Hakan Kjellerstrand, hakank@gmail.com
  See also my SWI Prolog page: http://www.hakank.org/swi_prolog/

*/

:- use_module(library(clpfd)).
:- use_module(hakank_utils).

% find all optimal solutions
go :-
        data(Price, Limits, [Calories, Chocolate, Sugar, Fat]),
        once(diet(Calories,Chocolate,Sugar,Fat,Price,Limits, Xs,XSum)),
        writeln([cost=XSum, Xs]),

        % and get all optimal solutions
        % (It happens to be a unique solution.)
        format("Find all solutions with cost ~d:\n", XSum),
        findall(Xs2,diet(Calories,Chocolate,Sugar,Fat,Price,Limits, Xs2,XSum),Sols),
        writeln(Sols),
        nl.


go2 :-
        data(Price, Limits, [Calories, Chocolate, Sugar, Fat]),       
        diet2([Calories,Chocolate,Sugar,Fat],Price,Limits, Xs,XSum),
        writeln([cost=XSum, Xs]).


% 
% Diet problem
%
diet(Calories,Chocolate,Sugar,Fat,Price,Limits, Xs,XSum) :-

        length(Price,Len),
        length(Xs, Len),
 	Xs ins 0..10,

        nth1(1,Limits,Limits1),
        nth1(2,Limits,Limits2),
        nth1(3,Limits,Limits3),
        nth1(4,Limits,Limits4),
        
        scalar_product(Calories,  Xs, #>=, Limits1), % 500,
        scalar_product(Chocolate, Xs, #>=, Limits2), %   6,
        scalar_product(Sugar,     Xs, #>=, Limits3), %  10,
        scalar_product(Fat,       Xs, #>=, Limits4), %   8,
        
        scalar_product(Price, Xs, #=, XSum), % to minimize
        
        % optimize or find all (optimal) solutions
        ( var(XSum)
        -> 
          labeling([min(XSum)], Xs)
        ; 
          % here XSum is bound so we just label the vars
          labeling([], Xs)
        ).

%
% This is a more general solution where all the nutritions 
% are handled in a foreach loop.
%
diet2(Products,Price,Limits, Xs,XSum) :-

        length(Price,Len),
        length(Xs,Len),
        Xs ins 0..10,

        scalar_product_rows(Products,Xs,#>=,Limits),
        scalar_product(Price, Xs, #=, XSum), % to minimize
        (var(XSum)
        ->
         labeling([min(XSum)], Xs)
        ; 
         labeling([],Xs)
        ).

scalar_product_rows([],_Xs,_Rel,_Limits).
scalar_product_rows([Product|Products],Xs,Rel,[Limit|Limits]) :-
        scalar_product(Product, Xs,Rel, Limit),
        scalar_product_rows(Products,Xs,Rel,Limits).

%
% data
%
data(Price, Limits, Nutrition) :-
        Price = [ 50, 20, 30, 80], % price in cents for each nutrition
        Limits = [500,  6, 10,  8], % limits, requirements for each nutrition type
        
        % nutrition for each product
        Nutrition = 
        [[400, 200, 150, 500],  % calories
         [  3,   2,   0,   0],  % chocolate
         [  2,   2,   4,   4],  % sugar
         [  2,   4,   1,   5]]. % fat


